File: alphaSwap.py
'''
Created by Aleksandar Damnjanovic
Date: 13.11.2021
Last edit: 22.11.2021
'''

from io import FileIO
import os
import docx2txt
import pdfplumber
import string

class alphaSwap:

    def __init__(self) -> None:
        self.d= self.parseCSV()

    def parseCSV(self):
        file= open('dictionary.csv', 'r')
        content= file.read()
        content=content.split('\n')
        _d= dict()
        for dd in content:
            dd= dd.split(',')
            _d[dd[0]]= dd[1]
        return _d

    def swapRawText(self, text):
        letter=""
        for c in text:
            try:
                letter+=str(self.d[c])
            except:
                letter+=str(c)
        return letter

    def swapTextFile(self, path):
        file= open(path, "r")
        text= file.read()
        file.close()
        t= self.swapRawText(text)
        fname= os.path.basename(path)
        nname= "swap_"+fname+".txt"
        fpath= str.replace(path, fname,"")
        npath=fpath+nname
        file= open(npath,"w")
        file.write(t)
        file.close()
        print("swap completed")
        
    def swapDocxFile(self, path):
        text= docx2txt.process(path)
        t= self.swapRawText(text)
        fname= os.path.basename(path)
        nname= "swap_"+fname+".txt"
        fpath= str.replace(path, fname,"")
        npath=fpath+nname
        file= open(npath,"w")
        file.write(t)
        file.close()
        print("swap completed")

    def swapPdfFile(self, path):
        text= pdfplumber.open(path)

        t=''
        for p in text.pages:
            t+= self.swapRawText(p.extract_text())

        fname= os.path.basename(path)
        nname= "swap_"+fname+".txt"
        fpath= str.replace(path, fname,"")
        npath=fpath+nname
        file= open(npath,"w")
        file.write(t)
        file.close()
        print("swap completed")

    def processDirectory(self, path):
        l=os.listdir(path)
        d=''
        for file in l:
            if d=='':
                d= str.replace(os.path.abspath(path), file, "")
                if str.endswith(d,"/")!=True:
                    d+=os.sep
                 
            if (d+file).endswith(".pdf"):
                self.swapPdfFile(d+file)
            elif (d+file).endswith(".docx"):
                self.swapDocxFile(d+file)
            elif (d+file).endswith(".txt"):
                self.swapTextFile(d+file)

File: alphaSwap.R
#*****************************
#Title: Alpha Swap
#Author: Aleksandar Damjanovic
#Date: 22.11.2021
#*****************************

library("readr")
library("officer")
library("pdftools")

alphaSwap_init<- function() {
  d<-read.csv('dictionary.csv')
  nm<-c('1','2')
  f<-data.frame('1'='A', '2'='A')
  colnames(d)<-nm
  colnames(f)<-nm
  dd<-rbind(f,d)
  return(dd)
}

parseRaw<-function(string){
  vec<-strsplit(string,"")
  vec<-as.vector(vec)
  vec1<-''
  for(i in 1:length(vec)){
    let<- dict$`2`[dict$`1`==vec[[1]][i]]
    if(length(let)=='NA'){
      let<- as.character(let)
      vec1<- paste(vec1, let, sep = "")
    }else{
      vec1<-paste(vec1, string[i], sep = "")
    }
  }
  vec1<-as.character(vec1)
  return(vec1)
}

parseTxt<-function(path){
  if(endsWith(path,'txt')!=TRUE){
    stop('txt extension is mandatory')
  }else{
    string<-read_file(path)
    content<-parseRaw(string)
    newPath<-paste('swap_', path)
    write_file(content, newPath, append = FALSE)
  }
}

parseWord<-function(path){
  dict<-alphaSwap_init()
  path<-'test.docx'
  content<-read_docx(path)
  content<-docx_summary(content)
  content<-content$text
  newPath<-paste('swap_', path, ".txt")
  if(file.exists(newPath))
    file.remove(newPath)
  for(i in 1:length(content)){
    t<- as.character(content[i])
    trans<-parseRaw(t)
    write_file(trans, newPath, append = TRUE)
    write_file('\n', newPath, append = TRUE)
  }
}

parsePdf<-function(path){
  pdf<-pdf_text(path)
  newPath<-paste('swap_', path, ".txt")
  if(file.exists(newPath))
    file.remove(newPath)
  for(i in 1:length(pdf)){
    trans<-parseRaw(pdf[i])
    write_file(trans, newPath, append = TRUE)
    write_file('\n', newPath, append = TRUE)
  }
}

parseDir<-function(path){
  path<-"./"
  
  files<-list.files(path,pattern = "*.txt")
  for(i in files){
    parseTxt(i)
  }
  
  files<-list.files(path,pattern = "*.docx")
  for(i in files){
    parseWord(i)
  }
  
  files<-list.files(path,pattern = "*.pdf")
  for(i in files){
    parsePdf(i)
  }
  
}